### PR TITLE
Improve security of communication with external components.

### PR DESCRIPTION
**Pull request by Creatone**

### CODE REVIEW

File: wca/security.py [hunk 1]
```diff
@@ -106,3 +106,32 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             os.seteuid(self.uid)
             log.log(logger.TRACE, "Effective user id from 0 to {}".format(self.uid))
             self.uid = 0
+
+
+class SSLCert():
+    cert_path: str
```
Comment (ppalucki) [inline wca/security.py]: use dataclasses and ``Path`` type

File: wca/security.py [hunk 2]
```diff
@@ -106,3 +106,32 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+    key_path: str
+
+    def __init__(self, cert_path: str, key_path: str):
+        assure_type(cert_path, str)
+        assure_type(key_path, str)
+
+        # Check if files are available
+        if not os.path.isfile(cert_path):
```
Comment (ppalucki) [inline wca/security.py]: there should be existing type of Path somewhere here

https://github.com/intel/workload-collocation-agent/blob/master/wca/config.py#L219

File: configs/mesos/mesos_external_detector.yaml [hunk 3]
```diff
@@ -3,6 +3,12 @@
 runner: !DetectionRunner
   node: !MesosNode
     mesos_agent_endpoint: 'http://127.0.0.1:5051'
+    ssl_verify: True
```
Comment (felidadae) [inline configs/mesos/mesos_external_detector.yaml]: please remove this line, as it is not needed anymore

File: wca/config.py [hunk 4]
```diff
@@ -215,10 +234,16 @@ def assure(cls, value):
         if cls.absolute:
             assure_absolute_path(value)
 
+        if cls.permission:
+            assure_permission_path(value, cls.permission)
+
+
+def Path(absolute: Optional[bool] = False,
+         max_size: Optional[int] = 400,
```
Comment (ppalucki) [inline wca/config.py]: Optional means that you can pass `None` not that it is kwarg

File: wca/config.py [hunk 5]
```diff
@@ -204,6 +213,16 @@ def Url(max_size=400, supported_schemes=('http', 'https'),
                          is_path_obligatory=is_path_obligatory))
 
 
+class Permission(int, Enum):
```
Comment (ppalucki) [inline wca/config.py]: drop it

File: wca/config.py [hunk 6]
```diff
@@ -215,10 +234,16 @@ def assure(cls, value):
...
+
+
+def Path(absolute: Optional[bool] = False,
+         max_size: Optional[int] = 400,
+         permission: Optional[Permission] = None):
 
-def Path(absolute=False, max_size=400):
     return _PathType('Path', (_PathType, SemanticType),
-                     dict(absolute=absolute, max_size=max_size))
+                     dict(absolute=absolute, max_size=max_size, permission=permission))
```
Comment (ppalucki) [inline wca/config.py]: replace permission with mode to match how it is used in os.access

File: wca/config.py [hunk 7]
```diff
@@ -215,10 +234,16 @@ def assure(cls, value):
         if cls.absolute:
             assure_absolute_path(value)
 
+        if cls.permission:
+            assure_permission_path(value, cls.permission)
+
+
+def Path(absolute: Optional[bool] = False,
+         max_size: Optional[int] = 400,
+         permission: Optional[Permission] = None):
```
Comment (ppalucki) [inline wca/config.py]: optional[int]

File: wca/security.py [hunk 8]
```diff
@@ -106,3 +110,43 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+    server_verify: Union[bool, Path(absolute=True, permission=Permission.READ)] = False
+    cert_path: Path(absolute=True, permission=Permission.READ) = None
+    key_path: Path(absolute=True, permission=Permission.READ) = None
+
+    def __post_init__(self):
+        assure_type(
+                self.server_verify,
+                Union[bool, Path(absolute=True, permission=Permission.READ)])
+
+        if bool(self.cert_path) != bool(self.key_path):
```
Comment (ppalucki) [inline wca/security.py]: give better comment above why casting to bool

File: wca/security.py [hunk 9]
```diff
@@ -106,3 +110,43 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            raise ValueError('Both certificate and key paths must be set!')
+
+        if bool(self.cert_path):
```
Comment (ppalucki) [inline wca/security.py]: use `is not None`

File: wca/security.py [hunk 10]
```diff
@@ -106,3 +110,43 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            assure_type(self.cert_path, Path(
```
Comment (ppalucki) [inline wca/security.py]: it is already checked [here](https://github.com/intel/workload-collocation-agent/blob/master/wca/config.py#L414)

File: wca/security.py [hunk 11]
```diff
@@ -106,3 +110,43 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             os.seteuid(self.uid)
             log.log(logger.TRACE, "Effective user id from 0 to {}".format(self.uid))
             self.uid = 0
+
+
+@dataclass
+class SSL():
+    server_verify: Union[bool, Path(absolute=True, permission=Permission.READ)] = False
+    cert_path: Path(absolute=True, permission=Permission.READ) = None
```
Comment (ppalucki) [inline wca/security.py]: add `Optional`

File: wca/mesos.py [hunk 12]
```diff
@@ -71,33 +71,31 @@ def find_cgroup(pid):
...
 
-    # A flag of python requests library to enable ssl_verify or pass CA bundle:
-    # https://github.com/kennethreitz/requests/blob/5c1f72e80a7d7ac129631ea5b0c34c7876bc6ed7/requests/api.py#L41
-    ssl_verify: Union[bool, Path] = True
-
     # Timeout to access mesos agent.
     timeout: Numeric(1, 60) = 5.  # [s]
 
+    # https://github.com/kennethreitz/requests/blob/5c1f72e80a7d7ac129631ea5b0c34c7876bc6ed7/requests/api.py#L41
+    ssl: Optional[SSL] = SSL()
```
Comment (ppalucki) [inline wca/mesos.py]: set default to None

File: wca/kubernetes.py [hunk 13]
```diff
@@ -77,21 +78,29 @@ class KubernetesNode(Node):
...
-        r = requests.get(full_url, json=dict(type='GET_STATE'),
-                         verify=False, cert=(self.client_cert, self.client_private_key))
-        r.raise_for_status()
 
-        return r.json()
+        with requests.get(
+                full_url,
+                json=dict(type='GET_STATE'),
+                timeout=self.timeout,
+                verify=self.ssl.server_verify,
```
Comment (ppalucki) [inline wca/kubernetes.py]: `verify = self.ssl.server_verify if self.ssl is not None else None`

File: wca/security.py (hunk 11 above)
Comment (ppalucki) [inline wca/security.py]: https://2.python-requests.org/en/master/user/advanced/#client-side-certificates

**Comment by ppalucki** [comment]
another problem:

```sh
(owca) ➜  owca (pr/149)  PYTHONPATH=. python wca/main.py -c $PWD/configs/kubernetes/kubernetes_example_allocator.yaml
2019-06-13 12:17:49,321 WARNING  {MainThread} [wca.main] This software is pre-production and should not be deployed to production servers.
2019-06-13 12:17:49,321 INFO     {MainThread} [wca.main] Version wca: 0.4.dev57+g48ce2a9
2019-06-13 12:17:49,324 ERROR    {MainThread} [wca.main] Error: Cannot load config file '/home/ppalucki/work/serenity/owca/configs/kubernetes/kubernetes_example_allocator.yaml': could not determine a constructor for the tag '!SSL'   in "/home/ppalucki/work/serenity/owca/configs/kubernetes/kubernetes_example_allocator.yaml", line 12, column 17. Available tags are: AllocationRunner, KafkaStorage, KubernetesNode, CgroupDriverType, LogStorage, AllocationConfiguration, NOPAllocator, NOPAnomalyDetector, StaticNode, StaticAllocator, DetectionRunner, MesosNode, MeasurementRunner
```

**Comment by ppalucki** [comment]
you have typos in kubernetes example, because field is now client_cert_path and example is just cert_path

File: wca/config.py [hunk 14]
```diff
@@ -104,6 +105,12 @@ def assure_absolute_path(value):
                               'setting it to False.')
 
 
+def assure_permission_path(value, permission):
+    if not os.access(value, permission):
+        raise PermissionError(
```
Comment (ppalucki) [inline wca/config.py]: replace that with ValidationError

**Pull request merged**